### PR TITLE
Made most `__device__` functions/methods `__host__ __device__`

### PR DESCRIPTION
**Pull request by kolayne**
~~Is blocked by #8. Current branch (`host-device_instead_of_device`) is based on `move_copy_mechanisms`, so the corresponding PR should be approved before approving this, and new commits at `move_copy_mechanisms` should be merged into this branch if there are any~~

I think it's wrong to make most functions/methods `__device__` only because if their parameters are allocated in host memory, they should be executed on host. This enhancement also makes it possible to perform some object setup while it's in host memory and then copy it to device memory (before it was only possible to create an object in device memory and set it up there). However, it's absolutely ok if some function used in setup or simulation uses some CUDA feature(s) and therefore is made `__device__` only (of course, it also requires all "parent" functions (which call that function) to be `__device__` either)

### CODE REVIEW

**Comment by kolayne** [comment]
I haven't edited `SimulationMap` class at all, because it has a `__global__` ~~method~~ friend function, so I'd say `SimulationMap` is designed to only exist in device memory, but probably this is wrong, I'm not sure

File: src/Particle.cuh [hunk 1]
```diff
@@ -105,6 +105,8 @@ public:
      *
      * @see Particle::capture
      */
+    // This function technically can be made `__host__ __device__`, but it is not made, to meet the declaration
+    // of `Polyhedron::capture`
```
Comment (tanya-kta) [inline src/Particle.cuh]: Suppose you meant `Particle::capture`
```suggestion
    // of `Particle::capture`
```

File: src/common.cuh [hunk 2]
```diff
@@ -50,7 +50,7 @@ __device__ T *malloc_and_copy(const T *const source, int count)
  * @returns Pointer to the created array
  */
 template<class T>
-__device__ T *device_realloc(T *source, int old_size, int new_size)
+__host__ __device__ T *device_realloc(T *source, int old_size, int new_size)
```
Comment (tanya-kta) [inline src/common.cuh:53]: `device_realloc` is now `__host__ __device__`, so `device` in the name of this function sounds strange. It relates to #4

**Review by tanya-kta — CHANGES_REQUESTED** [review]
Found a few issues

File: src/common.cuh (hunk 2 above)
Comment (kolayne) [reply src/common.cuh:53]: I think we can ignore it for now and leave it to fixers of #4

**Review by tanya-kta — APPROVED** [review]
OK

**Pull request merged**